projects/meta_cl/networks/dendritic_networks.py

```diff
@@ -107,3 +109,107 @@ def forward(self, x):
+                 num_segments=20,
+                 dim_context=100,
+                 dendrite_sparsity=0.70,
+                 **kwargs):
+
+        super().__init__()
+
+        self.segments = DendriteSegments(
+            num_units=2304,
+            num_segments=num_segments,
+            dim_context=dim_context,
+            sparsity=dendrite_sparsity,
+            bias=None,
+        )
+        self.classifier = nn.Linear(2304, num_classes)
+
+        self.prediction = nn.Sequential(
+            *self.conv_block(3, 256, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(256, 256, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(256, 256, 3, 1, 0),
+            nn.AdaptiveAvgPool2d(output_size=(3, 3)),
+            nn.Flatten(),
+        )
+
+        self.modulation = nn.Sequential(
+            *self.conv_block(3, 112, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(112, 112, 3, 1, 0),
+            nn.MaxPool2d(kernel_size=2, stride=2),
+            *self.conv_block(112, 112, 3, 1, 0),
+            nn.AdaptiveAvgPool2d(output_size=(3, 3)),
+            nn.Flatten(),
+            nn.Linear(1008, dim_context)
+        )
+
+        # Apply Kaiming initialization
+        self.reset_params()
+
+    def reset_params(self):
+        # Apply Kaiming initialization to Linear and Conv2d params
+        named_params = filter_params(self, include_modules=[nn.Linear, nn.Conv2d])
+        for _, param in named_params.items():
+            if param.ndim > 1:
+                nn.init.kaiming_normal_(param)
+            else:
+                nn.init.zeros_(param)
+
+    @classmethod
+    def conv_block(cls, in_channels, out_channels, kernel_size, stride, padding):
+        return [
+            nn.Conv2d(
+                in_channels=in_channels, out_channels=out_channels,
+                kernel_size=kernel_size, stride=stride, padding=padding
+            ),
+            nn.ReLU(),
+        ]
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """
+        Apply dendrites as a gating mechanism.
+
+        TODO: Decide best method to apply dendrites.
+        """
+        # # Multiple by the sigmoid of the max along each segment.
+        # return y * torch.sigmoid(dendrite_activations.max(dim=2).values)
+
+        inds = dendrite_activations.abs().max(dim=2).indices
+        inds = inds.unsqueeze(dim=2)
+        dendrite_activations = torch.gather(dendrite_activations, dim=2, index=inds)
+        dendrite_activations = dendrite_activations.squeeze(dim=2)
+        dendrite_activations = torch.sigmoid(dendrite_activations)
+        return y * dendrite_activations
+
+    def forward(self, x):
+        mod = self.modulation(x)
+        pred = self.prediction(x)
+
+        dendrite_activations = self.segments(mod)
+        out = self.apply_dendrites(pred, dendrite_activations)
```

Comment:
why not just use the `AbsoluteMaxGatingDendriticLayer` here, instead of redefining the `forward` and `apply_dendrites` methods?

Reply:
It's actually a different set of operations being performed here. The gating is being applied to the flattened output of a conv layer, and `AbsoluteMaxGatingDendriticLayer` doesn't do that nor does its 2d counterpart.

When we made the `DendriticSegments`, I had it in mind to be able to use it directly as needed whenever we didn't have a layer that neatly wrapped the desired functionality. As we play around with architectures, I'm sure we'll find some canonical form and we can make higher-level classes accordingly.